### PR TITLE
fix: use linear interpolation to implement range LINEAR fill strategy

### PR DESCRIPTION
**Pull request by Taylor-lagrange**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

Use linear interpolation to implement range `LINEAR` fill strategy

The previous `LINEAR` strategy of range fill was implemented using the average value. Only the data before and after an null data point are non-empty, and the value of this point will be filled with the average of the two numbers. Now the implementation is changed to linear interpolation. As long as there are more than two non-empty data points in the entire time series, all null data points can be filled in using linear interpolation.

## Checklist

- [x]  I have written the necessary rustdoc comments.
- [x]  I have added the necessary unit tests and integration tests.

### CODE REVIEW

File: src/query/src/range_select/plan.rs [hunk 1]
```diff
@@ -117,32 +120,88 @@ impl Fill {
...
-                                        return Err(DataFusionError::Execution(
-                                            "RangePlan: Apply Fill LINEAR strategy on Non-floating type".to_string()));
-                                    } else {
-                                        continue;
-                                    }
-                                }
-                            }
-                        }
-                    }
+                    Fill::Linear => {}
```
Comment (killme2008) [inline src/query/src/range_select/plan.rs:123]:
```suggestion
                    Fill::Linear => unreachable!()
```

File: src/query/src/range_select/plan.rs [hunk 2]
```diff
@@ -117,32 +120,88 @@ impl Fill {
...
+        // We need at least two non-empty data points to perform linear interpolation
+        if not_null_num < 2 {
+            return Ok(());
+        }
+        let mut index = 0;
+        while index < data.len() {
+            // find null interval [start, end)
+            let start = data[index..]
+                .iter()
+                .position(ScalarValue::is_null)
```
Comment (killme2008) [inline src/query/src/range_select/plan.rs:144]: Looks like it's not efficient. I think we can find all the null indexes before this iteration. And use this info to replace the `start` and `end` calculation.

File: src/query/src/range_select/plan.rs [hunk 3]
```diff
@@ -117,32 +120,88 @@ impl Fill {
...
+            (*y0 as f64, *y1 as f64, true)
+        }
+        _ => {
+            return Err(DataFusionError::Execution(
+                "RangePlan: Apply Fill LINEAR strategy on Non-floating type".to_string(),
+            ));
+        }
+    };
+    for i in start..end {
+        let val = y0 + (y1 - y0) / (x1 - x0) * (ts[i] as f64 - x0);
```
Comment (killme2008) [inline src/query/src/range_select/plan.rs:197]: May the dividend be zero? We can do defensive programming here.

**Review by killme2008 — COMMENTED** [review]
Good job!  Question: is this algorithm the same with other TSDB? Like the prometheus for example.

File: src/query/src/range_select/plan.rs (hunk 2 above)
Comment (Taylor-lagrange) [reply src/query/src/range_select/plan.rs:144]: Even if you find it out in advance, you still need such iteration, and pay extra memory overhead to store `start`/`end`. The whole implementation is `O(n)`. Each scan starts from the next interval where an interval may occur (from `index`, so no duplicate iteration), but we can indeed avoid secondary traversal by storing the head and tail null intervals.

Comment (killme2008) [reply src/query/src/range_select/plan.rs:144]: > we can indeed avoid secondary traversal by storing the head and tail null intervals.

That would be great!

File: src/query/src/range_select/plan.rs (hunk 3 above)
Comment (Taylor-lagrange) [reply src/query/src/range_select/plan.rs:208]: It can't happen in theory, but we can do that

**Review by killme2008 — APPROVED** [review]
LGTM

**Comment by Taylor-lagrange** [comment]
> Good job! Question: is this algorithm the same with other TSDB? Like the prometheus for example.

The idea mainly comes from questdb/influxdb, they all use linear interpolation as fill strategy, our original implementation is too simple.

File: src/query/src/range_select/plan.rs [hunk 4]
```diff
@@ -117,32 +120,99 @@ impl Fill {
...
-                                        return Err(DataFusionError::Execution(
-                                            "RangePlan: Apply Fill LINEAR strategy on Non-floating type".to_string()));
-                                    } else {
-                                        continue;
-                                    }
-                                }
-                            }
-                        }
-                    }
+                    Fill::Linear => unreachable!(),
```
Comment (waynexia) [inline src/query/src/range_select/plan.rs:123]: Add some document about this `unreachable!()`

File: src/query/src/range_select/plan.rs [hunk 5]
```diff
@@ -1339,29 +1400,76 @@ mod test {
...
-                .unwrap_err()
-                .to_string(),
-            "Execution error: RangePlan: Apply Fill LINEAR strategy on Non-floating type"
-        );
-        let mut test2 = vec![
-            ScalarValue::Float32(Some(8.0)),
+    }
+
+    #[test]
+    fn test_fill_linear() {
```
Comment (waynexia) [inline src/query/src/range_select/plan.rs:1415]: Can this handle changing slope? E.g., `1, x, 3, x, 9`

Comment (Taylor-lagrange) [reply src/query/src/range_select/plan.rs:1417]: Test has been updated

File: src/query/src/range_select/plan.rs (hunk 4 above)
Comment (Taylor-lagrange) [reply src/query/src/range_select/plan.rs:125]: done

File: src/query/src/range_select/plan.rs (hunk 5 above)
Comment (waynexia) [reply src/query/src/range_select/plan.rs:1417]: I think you should generate interpolated data with different slopes, not overwriting existing points. This is what InfluxQL does:

image

**Review by waynexia — APPROVED** [review]

**Pull request merged**